From ea2c7332c1bfd98233de46deab851e1e86640d9b Mon Sep 17 00:00:00 2001
From: Raviteja Sunkara <raviteja@codeaurora.org>
Date: Mon, 18 Feb 2013 01:46:28 +0530
Subject: [PATCH] add Condition structure to SubmitRecord.

This commits adds support for using new Term Structure
'condition' similar to 'label' in prolog submit rules.
'condition' term accepts only two states,
*ok* - satisfies the condition
*need* - does not satisfy the condition.

Using 'condition' term one can add submit rules for a project
which is not related to Approvals(for which 'label' is used)

Change-Id: I4398438f89721490370103fbc8433c9fd94886f0
---
 .../google/gerrit/common/data/SubmitRecord.java    |   22 +++++++
 .../gerrit/server/project/ChangeControl.java       |   59 +++++++++++++------
 2 files changed, 62 insertions(+), 19 deletions(-)

diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/data/SubmitRecord.java b/gerrit-common/src/main/java/com/google/gerrit/common/data/SubmitRecord.java
index 365f6a9..3447bf0 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/data/SubmitRecord.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/data/SubmitRecord.java
@@ -42,8 +42,23 @@ public class SubmitRecord {
 
   public Status status;
   public List<Label> labels;
+  public List<Condition> conditions;
   public String errorMessage;
 
+  public static class Condition {
+    public static enum Status {
+      OK,
+      NEED;
+    }
+    public String name;
+    public Status status;
+
+    @Override
+    public String toString() {
+      return name + ": " + status;
+    }
+  }
+
   public static class Label {
     public static enum Status {
       /**
@@ -111,6 +126,13 @@ public class SubmitRecord {
         delimiter = ", ";
       }
     }
+    if (conditions != null) {
+      String delimiter = "";
+      for (Condition c : conditions) {
+        sb.append(delimiter).append(c);
+        delimiter = ", ";
+      }
+    }
     sb.append(']');
     return sb.toString();
   }
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/project/ChangeControl.java b/gerrit-server/src/main/java/com/google/gerrit/server/project/ChangeControl.java
index d0185dd..b994d77 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/project/ChangeControl.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/project/ChangeControl.java
@@ -425,38 +425,59 @@ public class ChangeControl {
         return logInvalidResult(submitRule, submitRecord);
       }
 
+      rec.conditions = new ArrayList<SubmitRecord.Condition> (submitRecord.arity());
       rec.labels = new ArrayList<SubmitRecord.Label> (submitRecord.arity());
 
       for (Term state : ((StructureTerm) submitRecord).args()) {
-        if (!state.isStructure() || 2 != state.arity() || !"label".equals(state.name())) {
+        if (!state.isStructure() || 2 != state.arity() || (!"label".equals(state.name())
+            && !"condition".equals(state.name()))) {
           return logInvalidResult(submitRule, submitRecord);
         }
 
-        SubmitRecord.Label lbl = new SubmitRecord.Label();
-        rec.labels.add(lbl);
+        if ("label".equals(state.name())) {
+          SubmitRecord.Label lbl = new SubmitRecord.Label();
+          rec.labels.add(lbl);
 
-        lbl.label = state.arg(0).name();
-        Term status = state.arg(1);
+          lbl.label = state.arg(0).name();
+          Term status = state.arg(1);
 
-        if ("ok".equals(status.name())) {
-          lbl.status = SubmitRecord.Label.Status.OK;
-          appliedBy(lbl, status);
+          if ("ok".equals(status.name())) {
+            lbl.status = SubmitRecord.Label.Status.OK;
+            appliedBy(lbl, status);
 
-        } else if ("reject".equals(status.name())) {
-          lbl.status = SubmitRecord.Label.Status.REJECT;
-          appliedBy(lbl, status);
+          } else if ("reject".equals(status.name())) {
+            lbl.status = SubmitRecord.Label.Status.REJECT;
+            appliedBy(lbl, status);
 
-        } else if ("need".equals(status.name())) {
-          lbl.status = SubmitRecord.Label.Status.NEED;
+          } else if ("need".equals(status.name())) {
+            lbl.status = SubmitRecord.Label.Status.NEED;
 
-        } else if ("may".equals(status.name())) {
-          lbl.status = SubmitRecord.Label.Status.MAY;
+          } else if ("may".equals(status.name())) {
+            lbl.status = SubmitRecord.Label.Status.MAY;
 
-        } else if ("impossible".equals(status.name())) {
-          lbl.status = SubmitRecord.Label.Status.IMPOSSIBLE;
+          } else if ("impossible".equals(status.name())) {
+            lbl.status = SubmitRecord.Label.Status.IMPOSSIBLE;
 
-        } else {
-          return logInvalidResult(submitRule, submitRecord);
+          } else {
+            return logInvalidResult(submitRule, submitRecord);
+          }
+
+        } else if ("condition".equals(state.name())) {
+          SubmitRecord.Condition c = new SubmitRecord.Condition();
+          rec.conditions.add(c);
+
+          c.name = state.arg(0).name();
+          Term status = state.arg(1);
+
+          if ("ok".equals(status.name())) {
+            c.status = SubmitRecord.Condition.Status.OK;
+
+          } else if ("need".equals(status.name())) {
+            c.status = SubmitRecord.Condition.Status.NEED;
+
+          } else {
+            return logInvalidResult(submitRule, submitRecord);
+          }
         }
       }
 
-- 
1.7.8.3

